packages/utils/src/helpers/index.ts

```diff
@@ -160,3 +173,70 @@ export const Result = {
     return { ok: false as const, error };
   },
 };
+
+
+// Translates Borsh type to JS type
+type FromBorsh<T> =
+  T extends "u8" | "u16" | "u32" ? number :
+  T extends "u64" | "u128" | "u256" | "u512" ? BN :
+  T extends "string" ? string :
+  T extends { kind: "option", type: infer S } ? FromBorsh<S> :
+  T extends new(...args: infer _) => infer Res ? Res :
+  T extends readonly unknown[] ? Uint8Array :
+  unknown;
+
+// Gets first element of a tuple pair
+type Key<T> =
+  T extends readonly [infer Key, unknown] ? Key : never;
+
+// Gets second element of a tuple pair
+type Value<T> =
+  T extends readonly [unknown, infer Value] ? Value : never;
+
+/**
+ * Turns a Borsh schema value into a type with fields translated from
+ * Borsh types to JS types.
+ *
+ * Useful when deserializing for making sure the object passed by
+ * Borsh to a constructor is properly typed. Without this, there is no
+ * guarantee that the object the constructor expects will match what
+ * Borsh passes to it as an argument.
+ *
+ * Usage:
+ *
+ *     class TxMsgValue {
+ *
+ *       constructor(args: SchemaObject<typeof TxMsgSchema>) {  // note use of typeof
+ *         args.token      // : string
+ *         args.fee_amount // : BN
+ *         ...
+ *       }
+ *     }
+ *
+ *     // schema value must match the following form:
+ *     const TxMsgSchema = [
+ *       TxMsgValue,
+ *       {
+ *         kind: "struct",
+ *         fields: [
+ *           ["token", "string"],
+ *           ["fee_amount", "u64"],
+ *           ["gas_limit", "u64"],
+ *           ["chain_id", "string"],
+ *         ],
+ *       },
+ *     ] as const; // as const needed for typeof to deduce types correctly
+ */
+export type SchemaObject<T> =
```

Comment:
I can't read it, but works great :)

Reply:
Sorry, the type-level stuff is quite cryptic but I couldn't see a clearer way to implement it.